lightstep/recorder.py

```diff
@@ -1,21 +1,106 @@
-""" This code is used to report traces to LightStep servers.
-"""
+from .basictracer.recorder import SpanRecorder
+from .crouton import ttypes
+
+import util
+import logging
+import constants
+import pprint
+import sys
 
 import atexit
 import contextlib
 import jsonpickle
 import random
 from socket import error as socket_error
 import ssl
-import sys
 import threading
 import time
 import warnings
 
 from thrift import Thrift
+from . import constants, version as cruntime_version, util, connection as conn #???
 
-from .crouton import ttypes
-from . import constants, version as cruntime_version, util, connection as conn
+
+class Recorder(SpanRecorder):
+    """ SpanRecorder's job is record and report a BasicSpan.
+    """
+    def __init__(self, *args, **kwargs):
+        self.runtime = Runtime(*args, **kwargs)
+        self._runtime_guid = self.runtime._runtime.guid
+
+    def record_span(self, span):
+        now_micros = util._now_micros()
+        
+        spanRecord = ttypes.SpanRecord(
+            trace_guid=str(span.context.trace_id),
+            span_guid=str(span.context.span_id),
+            runtime_guid=str(span._tracer.recorder._runtime_guid),
+            span_name=str(span.operation_name),
+            join_ids=[],
+            oldest_micros=long(span.start_time),
+            youngest_micros = long(now_micros),
+            attributes=[],
+            log_records=[]
+        )
+
+        for key in span.tags:
+            if key[:5] == "join:":
+                spanRecord.join_ids.append(ttypes.TraceJoinId(key, span.tags[key]))
+            else:
+                spanRecord.attributes.append(ttypes.KeyValue(key, span.tags[key]))
+
+        for log in span.logs:
+            event = ""
+            if len(log.event)>0:
+                #Don't allow for arbitrarily long log messages.
+                if sys.getsizeof(log.event)>constants.MAX_LOG_MESSAGE_LEN:
+                    event = log.event[:constants.LONGEST_POSSIBLE_SUB_LOG]
+                else:
+                    event = log.event
+            spanRecord.log_records.append(ttypes.LogRecord(stable_name= event, payload_json= str(log.payload)))
```

Comment:
I was just looking at the old impl: https://github.com/lightstep/lightstep-tracer-python/pull/6/files#diff-8eccd37adb452e5c153705aa18b16514L165

It waits until we're off of the calling thread to deal with the payloads, and I believe that was because of latency concerns with payload serialization. The good news is that it looks like the `Runtime` code can handle all of this and just expects you to pass `payload_json=log.payload`... I think it's taken care of here: https://github.com/lightstep/lightstep-tracer-python/pull/6/files#diff-60eb910fdf2a6c86e5906627144ef8acL194

Make sense?